### ISSUE TITLE
[ANR] com.google.firebase.messaging.EnhancedIntentService.processIntent

### ISSUE THREAD

**Issue by fdspbsteam** [firebase/quickstart-unity#1291]
### [READ] For Firebase Unity SDK issues, please report to [Firebase Unity open-source](https://github.com/firebase/firebase-unity-sdk/issues/new/choose)

Once you've read this section and determined that your issue is appropriate for this repository, please delete this section.

### [REQUIRED] Please fill in the following fields:

Unity editor version: 2019.4.16f1
Firebase Unity SDK version: 8.9.0
Source you installed the SDK: .tgz packages from download archive
Problematic Firebase Component: FirebaseInstanceIdReceiver
Other Firebase Components in use: Auth, Analytics, Crashlytics, Messaging
Additional SDKs you are using: Adjust, Facebook
Platform you are using the Unity editor on: Mac
Platform you are targeting: iOS, Android
Scripting Runtime: IL2CPP

### [REQUIRED] Please describe the issue here:

![image](https://user-images.githubusercontent.com/77659032/176105567-2089a11d-b2d1-4dbe-92b7-549de56d05aa.png)
![image](https://user-images.githubusercontent.com/77659032/176105597-c052d9d3-c00d-4fa5-97c0-7e24aaa4df77.png)
![image](https://user-images.githubusercontent.com/77659032/176105640-05b3b86d-cfb8-4562-aed3-dd792648ad68.png)
![image](https://user-images.githubusercontent.com/77659032/176105695-f7436cfe-54ae-4b8c-bb81-bd970a5cf12b.png)

#### Steps to reproduce:

Have you been able to reproduce this issue with just the Firebase Unity quickstarts (this GitHub project)?
Regarding our user event, this ANR occurs when a user receives an event **notification receive**

#### Relevant Code:

```
// TODO(you): code here to reproduce the problem
```

**Comment by paulinon** [comment]
Hi @fdspbsteam,

Could you provide relevant code snippets that make use of the Messaging SDK or Firebase in general? Does the crash occur when the notification is received while the app is running or does it appear after tapping the notification?

**Comment by fdspbsteam** [comment]
@paulinon `FirebaseAnalytics.LogEvent(eventName);`
**Does the crash occur when the notification is received while the app is running or does it appear after tapping the notification?**
We cannot reproduce it locally, but we saw in our users events "notification event" and then the crash

**Comment by fdspbsteam** [comment]
@paulinon I`m not sure that I catch your question, could you rephrase please.  In my opinion, the root cause displayed on screenshots and the fix is easy.

**Comment by paulinon** [comment]
Hi @fdspbsteam,

We'll need all the information we can get in order to identify what's causing these crashes. If you can't locally reproduce this, we can start with the Firebase-related code you're using. Otherwise, we're just runing in circles here.

**Comment by paulinon** [comment]
Hi @fdspbsteam,

Is there a task that follows `SubscribeAsync`? Do you have a method named `OnTokenReceived()` or `OnMessageReceived` in your code? Your implementation would be easier to interpet if your provide entire functions that use Firebase rather than lines of code.

**Comment by fdspbsteam** [comment]
Hi @paulinon 
in our code, we do:

`    FirebaseMessaging.UnsubscribeAsync(topic);
                FirebaseMessaging.SubscribeAsync(topic);
                `
                
                we use `EventHandler<TokenReceivedEventArgs> TokenReceived` and `    public static event EventHandler<MessageReceivedEventArgs> MessageReceived`

**Comment by paulinon** [comment]
Hi @fdspbsteam,

Could you provide the full functions corresponding to your `OnNotificationMessageReceived` and `OnFirebaseTokenReceived` events?

**Comment by fdspbsteam** [comment]
> Hi @fdspbsteam,
> 
> Could you provide the full functions corresponding to your `OnNotificationMessageReceived` and `OnFirebaseTokenReceived` events?

`  private void OnNotificationMessageReceived(object sender, MessageReceivedEventArgs args)
    {
        FMLogger.vCore("received");

        if (args != null && args.Message != null)
        {
            string pushId = string.IsNullOrEmpty(args.Message.MessageId) ? "null" : args.Message.MessageId;
            AppState.PushNotificationId = pushId;

            var data = args.Message.Data;
            if (data != null)
            {
                foreach (var kv in data)
                {
                    FMLogger.vCore(string.Format("{0}:{1}", kv.Key, kv.Value));
                    if (kv.Key == APP_MESSAGE_KEY)
                    {
                        string queue = kv.Value;
                        var msg = UrlSchemeParser.TryParse(queue, AppMessageSource.PushNotification);
                        if (msg != null)
                        {
                            NotificationAppMessageReceived?.Invoke(msg, pushId);
                            return;
                        }
                    }
                }
            }
        }
        else
        {
            Debug.LogError("empty");
        }
    }`
    
    `    private void OnFirebaseTokenReceived(object sender, TokenReceivedEventArgs e)
    {
        FMLogger.vCore("token received. token:\n" + (e != null ? e.Token : "null"));
        string token = (e != null && !string.IsNullOrEmpty(e.Token)) ? e.Token : null;

#if UNITY_ANDROID
        if (token != null)
            AdjustHelper.SetUninstallToken(e.Token);
#elif UNITY_IOS
        TimeoutTaskRunner.Instance.Run(CheckForIOSToken());
#endif
        if (token != null)
            TokenReceived?.Invoke(token);
    }`

**Comment by paulinon** [comment]
Thanks for the additional information, @fdspbsteam. I've relayed this to the team so that they could take a look at this issue. With that, I'll be marking this as a bug for now.

**Comment by felipecastrosales** [comment]
any updates in this?

**Comment by FangJZSP** [comment]
Are there any new developments?
